lib/searches.js

```diff
@@ -1,252 +1,363 @@
+let getBFSLinks = function(url, level, sourceURL, sourceTitle, searchDepth, searchID, keyword, result) {
+
+    // Make request
+    let fetch = request({
+        url: url,
+        timeout: 30000,
+        jar: true
+    });
+    return fetch.then((body) => {
+        try {
+            // Load HTML and get all links in page
+            let title = body.match(TITLE);
+            if (!title) { // response did not contain a <title>
+                title = "";
+            } else {
+                title = title[1]; // get first matching group
+                if (title.length > 200) {
+                    title = "";
+                }
+            }
+            console.log("Link <title>", title);
+
+            // Insert link into temp edges table
+            res = dbAPI.insertIntoTempEdges(searchID, sourceURL, sourceTitle, url, title).catch((err) => {
+                console.log(err);
+            });
+
+            // Find keyword if present
+            if ((keyword !== "") && (body.indexOf(keyword) > -1)) {
+                // We found the search term: we're done.
+                result.status = TERM_FOUND;
+                result.keywordURL = url;
+                return [];
+            }
+          
+            // If URL is not a leaf node, get links
+            if (level >= searchDepth) {
+                body = null;
+                return [];
+            }
+            
+            let numLinks = 0;
+            let linkList = [];
+            let links = body.match(ALL_LINKS) || [];
+
+            links.forEach((link) => {
+                let linkURL = link.match(ACTUAL_LINK);
+                if (!linkURL) {
+                    return [];
+                }
+                linkURL = linkURL[1];
                 
-                 // Insert link into temp edges table
-                let tempInsert = dbAPI.insertIntoTempEdges(url, linkUrl, cookie, 'Ian Dalrymple', 'BFS', searchDepth, rootUrl);   
-                tempInsert.catch((err) =>  {
-                    throw err;
-                });
-               
-                // Add links that need to be crawled to toCrawl array
-                if (level < searchDepth) {
-                    toCrawl.push({
-                        url: linkUrl,
-                        level: level
+                // Try to resolve relative links with host name
+                if (linkURL && !VALID_URL.test(linkURL)){
+                    linkURL = resolveUrl.resolve(url, linkURL);
+                }
+                
+                // If link looks valid, add to link list
+                if (linkURL && validUrl.isUri(linkURL)) {
+                    linkList.push({
+                        "SourceURL": url,
+                        "SourceTitle": title,
+                        "DestinationURL": linkURL,
                     });
                 }
+                
+                numLinks++;
+                if (numLinks >= THRESHOLD) {
```

Comment:
This needs to move to the top of the loop, the `return` only moves to the next iteration of the `forEach`. In the getDFSLinks version we `return false`, which _will_ stop that loop since it's a cheerio "each" and not an array "forEach". If this is at the top then all the rest of the links will bail out early even when they're called.

That said, I was able to get through a BFS of depth 2 on flip with this, so maybe there were some other changes which fixed the problems? Let's move this to the top and change the `THRESHOLD` variable you added to 1000 or something and see if it still works.